Add niconico video Provider

Shi-553 opened the pull request:
This PR adds a new provider that enables searching, playback, and library integration for music content on niconico video ([nicovideo.jp](https://www.nicovideo.jp)) within Music Assistant.

## Highlights

- Search & Explore
  - Keyword search for Tracks, Playlists (Mylist), and Albums (Series)
  - Recommendations based on history, follow activity, like history
  - Fetch similar tracks
- Library
  - Playlists: your own Mylist (read/write)
  - Artists: your following artists
- Streaming Playback
  - Uses StreamType.CUSTOM with dynamic m3u8 generation for optimized seeking
  - HLS playlist is parsed once during stream conversion and reused for all seek operations
  - Generates dynamic m3u8 starting from the target segment for fast seek response

## Authentication

- Prefer `user_session`(cookie); on failure, fall back to Email/Password (+MFA) [[auth.py]](https://github.com/music-assistant/server/pull/2339/files#diff-513394567dfe2aa6cfab258602f18ca42f6e61b040a7c849835210f78d39ecaa)
- Persist session upon success and schedule re-login every 30 days

## Dependencies
- Uses [niconico.py](https://github.com/niconicolibs/niconico.py)
- An upstream PR has been submitted and is pending merge; currently pinned to a commit in the [fork](https://github.com/Shi-553/niconico.py)
- After merge, we plan to switch to the stable PyPI release
  - Should I wait until the merge before submitting a pull request?

## Tests and Fixtures

- Added snapshot tests for the converter. [[test_converters.py]](https://github.com/music-assistant/server/pull/2339/files#diff-8fe21ab9bb121772d32efc3cf884595571de1f222a279e5370fe58cb7af06d00)

## Known Issues

- If you assign an album to a track, the track-specific thumbnail will not be used.
  - Is it necessary to add a flag variable to the Track model? [[track.py]](https://github.com/Shi-553/server/blob/add-niconico-provider/music_assistant/providers/nicovideo/converters/track.py)

Commit:
Initial comit of NicoNico provider

Commit:
Remove unwanted logout calls

Commit:
niconico: Split and organize classes

Commit:
Add artist retrieval method and fix Explorer mixin imports in Niconico provider

Commit:
Niconico: Enhance artist parsing with metadata and SNS links

Commit:
Niconico: Add sensitive content handling and improve recommendations feature

Commit:
Niconico: Implement series handling as albums and enhance library retrieval methods

Commit:
Niconico: Refactor the structure of the Nikoniko adapter and split the individual adapters into separate files.

Commit:
Niconico: Add URLs for playlists, tracks, and artists in parsers; update import in album mixin

Commit:
Refactor Niconico provider mixins and add configuration management

- Enhanced NiconicoMusicProviderExplorerMixin to fetch recommendations with filtering based on user history, following activities, and like history.
- Introduced NiconicoConfig class for managing provider-specific configurations, including recommendation counts and tag filtering.
- Implemented TagManager for caching and retrieving video tags asynchronously, with support for deduplication and periodic cleanup.
- Updated library mixin to support adding and removing tracks with auto-like functionality.
- Improved playlist mixin to handle adding and removing tracks from playlists with error handling.
- Added support for fetching and filtering tracks based on required tags in the track mixin.
- Created a new configuration file for Niconico provider settings.

Commit:
Niconico: Update workspace configuration to include additional paths and launch settings

Commit:
Niconico: Improve descriptions in configuration entries for clarity and consistency

Commit:
Niconico: Simplify get_artist_toptracks method to retrieve the newest 50 tracks

Commit:
Niconico: Add auto-sync and include library track artists settings; refactor library mixin

Commit:
Niconico: Refactor search methods for improved clarity and efficiency; update configuration entries for email and sensitive content handling

Commit:
Refactor Niconico search and user adapters; update parsing methods for improved track retrieval and error handling

Commit:
Niconico: Add methods for retrieving own videos and series; update configuration entries for improved content management

Commit:
Niconico: Enhance search and recommendation features; refactor configuration and constants for improved tag handling

Commit:
Refactor video parsing methods to handle None values and skip muted videos for improved track retrieval

Commit:
Remove Niconico server workspace configuration file

Commit:
Refactor Niconico provider: Enhance error handling, logging

- Replaced call_with_throttler with _call_with_throttler in NiconicoVideoAdapter for consistency.
- Removed handle_niconico_errors context manager and replaced it with direct logging in various mixins.
- Introduced log_verbose and log_verbose_operation functions for improved logging.
- Updated methods across mixins to utilize new logging functions and streamline error handling.
- Refactored tag management to improve caching and fetching logic.

Commit:
Niconico: Update icon SVG files for improved styling and add monochrome version

Commit:
Niconico: Renamed parsing to conversion functions for playlists, tracks, and series

Commit:
Refactor Niconico video adapter and track mixin: Improve type hints, enhance error handling, and streamline stream format processing

Commit:
Refactor type hints in adapter

Commit:
Renamed the provider from “niconico” to “nicovideo”.

Commit:
Nicovideo: Enhance cache management by limiting concurrent operations to reduce database load

Commit:
Nicovideo: Include own uploaded videos in library track retrieval

Commit:
Remove unused sensitive contents configuration from Nicovideo provider

Commit:
Update NicoVideo monochrome icon SVG with new design and filter effects

Commit:
Nicovideo: Add icons for various sections in the music provider explorer mixin

Commit:
Nicovideo: Improve tag fetching with async task locking to prevent race conditions

Commit:
Merge branch 'dev' into add-niconico-provider

Commit:
Nicovideo: Refactor provider mixins and enhance core functionality with MixinCaller

Commit:
Nicovideo: Adjust low priority throttler period for improved background tag updates

Commit:
Refactor Nicovideo provider: enhance video fetching and tag management, remove TagManager

Commit:
Refactor logging in Nicovideo provider: Adjusted log levels.

Commit:
Nicovideo: Enhance error logging for ValidationError in API calls

Commit:
Refactor Nicovideo provider mixins: streamline library item retrieval and enhance caching logic

Commit:
Refactor Nicovideo Adapter Structure and Update Mixins

- Introduced a new NicovideoAdapterHub to centralize adapter functionality.
- Updated mixins to use the new adapter hub instead of individual adapters.
- Implemented utility functions for common operations across converters.
- Improved metadata handling and mapping for media items.
- Ensured compatibility with the latest niconico API changes.

Commit:
Refactor NicovideoTrackConverter: Move audio format creation logic to a dedicated method and remove unused utility function

Commit:
Refactor Nicovideo provider mixins: rename mixin files.

Commit:
Renamed from adapter to service.

Commit:
Refactor Nicovideo service architecture: rename Hub to Manager

Commit:
Add override decorator to get_stream_details method in NicovideoMusicProvider

Commit:
Add override decorators to library methods in artist and track mixins

Commit:
Refactor Nicovideo converters: replace utility functions with helper methods

Commit:
Remove unused logging import from album mixin

Commit:
First implementation of Nicovideo testing.

Commit:
Refactor fixture data generation and saving process

- Updated JSON fixture files with dummy descriptions for testing purposes.
- Enhanced helper functions for consistent snapshot comparison and dynamic field stabilization.
- Introduced new scripts for generating and saving fixture data with integrated diff tracking.
- Improved logging for fixture generation and changes.

Commit:
feat: update yt-dlp version to 2025.8.11 in manifest and requirements

Commit:
feat: add DUMMY_IS_PEAK_TIME constant and update helper functions to use it; update requirements to include yt-dlp

Commit:
Merge remote-tracking branch 'upstream/dev' into add-niconico-provider

Commit:
Refactor track handling in NicovideoMusicProviderTrackMixin to improve stream metadata retrieval

Commit:
requirements: add yt-dlp version 2025.6.30 to dependencies

Commit:
fix: update artist mapping to return UniqueList for better handling

Commit:
feat: update workflow and main.py to latest versions from dev and fixture-updates branches

Commit:
fix: update schedule comment to reflect daily execution and enhance condition for fixture updates

Commit:
fix: update descriptions in test snapshots to use dummy text for consistency

Shi-553 committed:
Merge branch 'music-assistant:dev' into add-niconico-provider

Commit:
fix: remove obsolete check-test-session script

Commit:
fix: update release dates and dummy data in test fixtures for consistency

Commit:
fix: enhance stabilization logic by consolidating dynamic field and count value processing

Commit:
fix: streamline provider unload logic and improve error handling

Commit:
Refactoring: refactor niconico service methods to unify series and mylist retrieval logic.Made niconico_py_client accessible via properties in the service.

Shi-553 committed:
Update fixtures-update-and-test.yml

Shi-553 committed:
Merge branch 'music-assistant:dev' into add-niconico-provider

Commit:
Refactor: streamline audio format handling by introducing create_audio_format function and centralizing constants

Commit:
Refactor: improve stream details handling by safely retrieving HTTP headers and simplifying audio format parameters

Commit:
update snapshot

Commit:
Fix: Add “extra_input_args” from streamdetails to get_preview_stream and adjust additional input arguments.

Shi-553 committed:
Update fixtures-update-and-test.yml

Commit:
Refactoring: Added a stream converter to enhance the nicovideo provider and improve stream detail processing. Removed dependency on yt-dlp.

Commit:
- add StabilizationInfo
- Add Stream Test
- Called ruff after automatic generation.

Commit:
Fix: Update comment to clarify usage of StreamConversionData serialization in import statement

Commit:
Fix: Update media type checks in artist and track mixins for consistency

Commit:
Fix: Update search_videos_by_tags method to accept a single tag and streamline video search logic

Commit:
Refactoring: Remove unused BaseModel serialization strategy and related imports from helpers.py

Commit:
Fix: Rename search_videos_by_tags method to search_videos_by_tag for consistency

Commit:
Fix: Update error messages in artist mixin for clarity and detail; add note on playlist duplicate entries

Commit:
Fix: Enhance logging with safe argument summaries and caller information extraction

Commit:
Fix: Update niconico.py requirement to specific commit for stability

Commit:
Fix: Remove follow/unfollow artists functionality from Nicovideo configuration and artist mixin

Commit:
Fix: Removed logout upon unloading, and prevented attempts to log in again when authentication information is not available.

Commit:
Fix: Remove unnecessary login checks in various mixins and services to streamline data retrieval

Commit:
Fix: Enhance album mapping creation with optional thumbnail URL and update artist yield logic to handle ItemMapping conversion

Commit:
Fix: Changed to directly generate album and artist mappings.

Commit:
Fix: Changed to directly generate album and artist mappings.

Commit:
Merge branch 'add-niconico-provider' of github.com:Shi-553/server into add-niconico-provider

Commit:
Fix: Update label for user session configuration to clarify cookie usage

Commit:
Fix: Remove stage from niconico video manifest

Commit:
Fix: Update workflow description to clarify NICONICO_SESSION requirement

Commit:
Fix: Rename nicovideo provider fixtures update and test workflow

Commit:
Fix: Simplify docstring in get_config_entries and add category comments in config entries

Commit:
Fix: Refactor get_config_entries_impl to separate config entry categories and simplify parameters

Commit:
Move :nicovideo workflows

Commit:
Fix: Update conversion function for SeriesData to handle album tracks

Commit:
Fix: Update usage instructions for fixture generation script to reflect correct environment variable

Commit:
Fix: Remove unused convert_to_netscape function and related imports

Commit:
Fix: Refactor configuration constants for nicovideo provider by moving keys to config.py and cleaning up imports

Commit:
Fix: Remove unused DUMMY_DESCRIPTION constant and update stabilization rule for description

Commit:
Refactor Nicovideo provider configuration system to use category-based structure

- Introduced a new configuration system for the Nicovideo provider, organizing settings into categories: Auth, Content, and Recommendations.
- Updated all relevant code to access configuration values through the new structure, replacing old method calls with direct attribute access.
- Enhanced clarity and maintainability of the configuration management by encapsulating related settings within dedicated classes.

Commit:
Fix: Move recommendation_filter_tags configuration to the correct position in RecommendationsConfigCategory

Commit:
Fix: Update recommendation configuration keys and labels for clarity and consistency

Commit:
Fix: Add integrated loudness to stream conversion output

Shi-553 committed:
Merge branch 'music-assistant:dev' into add-niconico-provider

Commit:
update snapshot loudness

Commit:
Add: Added dependency on niconico.py for lint testing.

Commit:
Fix: Update niconico.py dependency PEP 508 format in manifest and requirements files

marcelveldt commented on .pre-commit-config.yaml:
```diff
@@ -66,6 +66,12 @@ repos:
+        language: system
+        types: [python]
+        entry: scripts/run-in-env.sh tests/providers/nicovideo/scripts/check-test-session.sh
+        files: ^tests/providers/nicovideo/.*\.py$
```
Shouldn't this belong to a client library ?

Shi-553 replied:
Sorry for the lack of context.
This script is specifically intended to prevent the TEST_USER_SESSION in tests/providers/nicovideo/fixtures/scripts/main.py from being accidentally committed.
I’ll make that more clear.

marcelveldt replied:
Let's find another way to prevent that - a provider should be isolated into its own

Shi-553 replied:
I've removed the `check-nicovideo-test-session` hook and implemented an alternative solution in commit c23c4aad:

**Changes:**
- ❌ Removed provider-specific pre-commit hook and check script
- ✅ Made `NICONICO_SESSION` environment variable mandatory (no hardcoded fallback)
- ✅ Added clear error message when the variable is not set

**Before:**
```python
# Allowed hardcoded session with "DO NOT COMMIT" comment
TEST_USER_SESSION = ""  # DO NOT COMMIT
session = TEST_USER_SESSION if TEST_USER_SESSION else os.getenv("NICONICO_SESSION")
```

**After:**
```python
# Only environment variable, fails immediately if not set
session = os.getenv("NICONICO_SESSION")
if not session:
    raise ValueError(
        "NICONICO_SESSION environment variable is required.\n"
        "Set it before running this script:\n"
        "  export NICONICO_SESSION='your_session_token'"
    )
```

This approach prevents accidental commits of hardcoded credentials while maintaining provider isolation - no changes to repository-wide configuration needed.

marcelveldt commented on music_assistant/providers/nicovideo/config/categories/auth.py:
```diff
@@ -0,0 +1,59 @@
+from music_assistant.providers.nicovideo.config.factory import ConfigFactory
+
+
+class AuthConfigCategory(ConfigCategoryBase):
+    """Authentication settings category."""
+
+    _auth = ConfigFactory("Authentication")
+
+    mail = _auth.str_config(
+        key="mail",
+        label="Email",
+        default=None,
+        description="Your NicoNico account email address.",
+    )
+
+    password = _auth.secure_str_or_none_config(
+        key="password",
+        label="Password",
+        description="Your NicoNico account password.",
+    )
+
```
So, are your config entries really that large complex that you need this multi layered pattern of providing the config entries ? Can you give some insights about the number of config entries ?

I would always start with a small/simple setup/PR and then extend later when/if needed.

Shi-553 replied:
Right.
The core part of the configuration is AuthConfigCategory.
I’ll adjust this PR to only include that minimal setup, and move the other categories into a separate PR later.

Shi-553 replied:
I've simplified the configuration as requested. The Content and Recommendations categories have been removed, leaving only the essential Authentication category (email/password).

Changes made in commits 85fe0eb4 and ed143c92:
- Removed `ContentConfigCategory` and `RecommendationsConfigCategory`
- Replaced config references with sensible defaults:
  - Include own mylists: `true`
  - Auto-like tracks: `true` 
  - Sensitive content handling: `"mask"`
  - Recommendation counts: 25/50/30

The complex configuration has been preserved in a separate branch (`add-nicovideo-advanced-options`) for potential future enhancement after this initial PR is merged.

This makes the provider much simpler to understand and maintain for initial users.

marcelveldt commented on music_assistant/providers/nicovideo/manifest.json:
```diff
@@ -0,0 +1,11 @@
+  "description": "Support for niconico video(nicovideo) in Music Assistant",
+  "codeowners": ["@Shi-553"],
+  "requirements": [
+    "niconico.py @ git+https://github.com/Shi-553/niconico.py@50a30c1bac30f657318ce1b60e65e2c6f002d094"
```
this is temporary accepted while the PR is in draft but when ready, it needs to be a pypi published package.

Shi-553 replied:
Regarding PyPI publication:

I've submitted pull requests ([#60](https://github.com/niconicolibs/niconico.py/pull/60)-[#63](https://github.com/niconicolibs/niconico.py/pull/63)) to the upstream `niconico.py` library back in August to contribute the features needed for this provider. However, there has been no response or review activity for about 2 months.


**My question:**
If the upstream PRs remain unreviewed, would it be acceptable to publish my fork to PyPI under a different package name (e.g., `niconico.py-ma`)? This would allow us to proceed with this Music Assistant integration while still giving the upstream maintainers time to review the changes.

Once/if the upstream PRs are merged and published, we could switch back to the official package name.

marcelveldt commented on music_assistant/providers/nicovideo/provider_mixins/__init__.py:
```diff
@@ -0,0 +1,28 @@
+"""
+nicovideo provider mixins package.
+
+Provider Mixins Layer: Business logic
+Implements Music Assistant provider interface methods.
+Each mixin handles specific media types and provider capabilities.
+"""
+
+from __future__ import annotations
+
+from .album import NicovideoMusicProviderAlbumMixin
+from .artist import NicovideoMusicProviderArtistMixin
+from .core import NicovideoMusicProviderCoreMixin
+from .explorer import NicovideoMusicProviderExplorerMixin
+from .playlist import NicovideoMusicProviderPlaylistMixin
+from .track import NicovideoMusicProviderTrackMixin
+
+# Defines the inheritance order for the NicovideoMusicProvider mixins.
+NICOVIDEO_MIXINS = (
+    NicovideoMusicProviderCoreMixin,
+    NicovideoMusicProviderTrackMixin,
+    NicovideoMusicProviderPlaylistMixin,
+    NicovideoMusicProviderArtistMixin,
+    NicovideoMusicProviderAlbumMixin,
+    NicovideoMusicProviderExplorerMixin,
+)
+
+__all__ = [cls.__name__ for cls in NICOVIDEO_MIXINS]
```
Be aware that by making your code more dynamic, it also makes it harder to read/follow by others. That might be an issue for any contributors that want to help out. Often simplicity is better than complexity, even if it results in more duplicate code.

If this is something you are fine with and you will be the core maintainer this is all fine.
Make sure that you try to have a library in between that does as much shaping of the data as possible so within MA you only need to call the methods and transform the data into MA models.

Shi-553 replied:
Just to confirm:
Is your comment about the provider being split into multiple mixins, or specifically about the dynamic generation of `__all__`?

For the “library in between” part: in my current design, that role is handled by the services/ (API coordination, rate limiting) and converters/ (transforming Niconico objects to MA models) layers.
Therefore, providers/mixins do not manipulate objects in niconico.py.

marcelveldt replied:
The comment was about both

Shi-553 replied:
I've addressed your feedback about code complexity. Commit 4ca29bd0 simplifies the mixin architecture:

**Changes made:**
- ✅ Removed `MixinCaller` class with complex lambda-based invoke methods
- ✅ Replaced with explicit `for` loops in provider methods
- ✅ Made class inheritance explicit - all 6 mixin classes listed directly in the class definition
- ✅ Converted `provider_mixins/__init__.py` to static `__all__` list
- ✅ Moved `NICOVIDEO_MIXINS` constant from `__init__.py` to `provider.py`

**Before (complex):**
```python
# Dynamic inheritance with type: ignore
class NicovideoMusicProvider(*NICOVIDEO_MIXINS):  # type: ignore[misc]
    
    # Lambda-based method invocation
    async def handle_async_init(self) -> None:
        await MixinCaller(self).invoke_all(
            lambda mixin: mixin.handle_async_init_for_mixin()
        )
```

**After (simple):**
```python
# Explicit inheritance
class NicovideoMusicProvider(
    NicovideoMusicProviderCoreMixin,
    NicovideoMusicProviderTrackMixin,
    NicovideoMusicProviderPlaylistMixin,
    NicovideoMusicProviderArtistMixin,
    NicovideoMusicProviderAlbumMixin,
    NicovideoMusicProviderExplorerMixin,
):
    
    # Simple for loop
    async def handle_async_init(self) -> None:
        for mixin_class in NICOVIDEO_MIXINS:
            await mixin_class.handle_async_init_for_mixin(self)
```

**Why I kept the mixin structure:**

The mixin-based separation is maintained for the following reasons:

1. **Clear separation of concerns**: Each mixin handles a specific media type (Track, Playlist, Artist, Album) or capability (Explorer, Core). This makes it easy to locate and modify functionality for a particular feature.

2. **File size management**: Without mixins, `provider.py` would be ~1500+ lines. Splitting into 6 focused files (~200-300 lines each) makes the codebase more navigable.

If you still think the mixin structure should be removed entirely, please let me know!

Commit:
Merge upstream/dev: resolve conflicts in pyproject.toml and requirements_all.txt

OzGav commented:
@Shi-553 just checking as to whether you have made any more progress with this one?

Shi-553 commented:
> just checking as to whether you have made any more progress with this one?

Sorry, I haven’t had a chance to work on this yet. I’ll try to make some time over the weekend.

Commit:
Fix nicovideo provider compatibility with upstream/dev API changes

- Remove CACHE_CATEGORY_MUSIC_PROVIDER_ITEM constant (no longer exists)
- Update cache.set() calls: replace base_key with provider parameter
- Remove category parameter from cache operations (defaults to 0)
- Convert _supported_features from ClassVar to instance method pattern
- Define SUPPORTED_FEATURES statically in __init__.py and pass to Provider
- Remove supported_features property override from NicovideoMusicProvider

Commit:
Remove unused get_supported_features_for_mixin methods

- Remove get_supported_features_for_mixin() from all mixins (base, artist, track, album, playlist, explorer)
- Remove unused ProviderFeature imports
- Features are now statically defined in __init__.py's SUPPORTED_FEATURES

Commit:
Simplify config: Remove Content and Recommendations categories

This change addresses PR review feedback requesting a simpler configuration.
The complex Content and Recommendations config categories have been removed
and replaced with sensible defaults:

- Content defaults: Include own mylists, exclude followed mylists
- Recommendations defaults: 25/50/30 track counts, no tag filtering
- Auto-like: Enabled by default
- Sensitive content: Mask by default

The removed complex configuration has been preserved in the
`add-nicovideo-advanced-options` branch for potential future enhancement.

This makes the initial provider setup much simpler while maintaining
core functionality.

Commit:
Refactor: Add @override, remove unused filtering, use SENSITIVE_CONTENTS constant

- Add missing @override decorator to get_library_playlists
- Remove _fetch_similar_tracks_with_filtering (no longer filters anything)
- Extract SENSITIVE_CONTENTS constant to nicovideo/constants.py
- Replace all hardcoded sensitive_contents with constant reference

These changes improve code clarity and maintainability.

Commit:
Update niconico.py dependency to the latest commit

Shi-553 reviewed (commented):
I've simplified the configuration as requested. The Content and Recommendations categories have been removed, leaving only the essential Authentication category (email/password).

Changes made in commits 85fe0eb4 and ed143c92:
- Removed `ContentConfigCategory` and `RecommendationsConfigCategory`
- Replaced config references with sensible defaults:
  - Include own mylists: `true`
  - Auto-like tracks: `true` 
  - Sensitive content handling: `"mask"`
  - Recommendation counts: 25/50/30

The complex configuration has been preserved in a separate branch (`add-nicovideo-advanced-options`) for potential future enhancement after this initial PR is merged.

This makes the provider much simpler to understand and maintain for initial users.

Commit:
refactor: Simplify mixin architecture - remove MixinCaller abstraction

- Remove MixinCaller class with complex lambda-based invoke methods
- Replace with explicit for loops in provider methods
- Make class inheritance explicit (list 6 mixin classes directly)
- Convert provider_mixins/__init__.py to static __all__ list
- Move NICOVIDEO_MIXINS constant to provider.py
- Change reversed() to [::-1] for type checker compatibility

This addresses PR feedback about code complexity while maintaining
mixin-based separation of concerns.

Commit:
refactor: Remove provider-specific pre-commit hook

Remove check-nicovideo-test-session hook to maintain provider isolation.
Instead, enforce environment variable usage in fixture generation script:

- Require NICONICO_SESSION environment variable (no hardcoded fallback)
- Add clear error message when variable is not set
- Update documentation to reflect environment-only authentication

This prevents accidental credential commits while keeping the provider
isolated from repository-wide configuration.

Commit:
test: Update converter snapshots for new model fields

Update test snapshots to include new fields added to Music Assistant models:
- grouping: Album grouping field
- in_library: Library membership flag

Shi-553 commented:
## Question about Track Thumbnail Behavior (From Known Issues)

For video platforms like Niconico, each track (video) has its own unique thumbnail that 
should be displayed. However, the current architecture always prefers album images:

- Backend: `controllers/media/base.py:876` and `albums.py:263`
- Frontend: `MediaItemThumb.vue:121`

Would it make sense to add a flag like `prefer_track_thumbnail: bool` to the Track model?
Or is there a design reason for always prioritizing album images?

Commit:
fix: nicovideo fixtures workflow - use pip instead of uv

- Remove dependency on uv and setup.sh
- Use standard pip installation like test.yml
- Remove unnecessary venv caching
- Simplify workflow by using GitHub Actions default Python environment

Commit:
fix: update niconico.py commit hash to latest (711729d)

- Align pyproject.toml with requirements_all.txt
- Fix dependency conflict in GitHub Actions

Commit:
refactor: remove nicovideo from mypy exclude list

- nicovideo provider is fully mypy strict mode compliant
- Dependencies managed via manifest.json and requirements_all.txt
- Align with other strict-mode compliant providers (spotify, deezer, etc.)

Commit:
fix: resolve mypy strict mode errors in nicovideo provider

- Add proper TYPE_CHECKING import for pydantic BaseModel
- Add explicit type casts to resolve no-any-return errors
- Import Any type for type annotations in tests
- Ensure all return statements have explicit types

Commit:
fix: add niconico.py to test dependencies for proper mypy type checking

- Add niconico.py to project.optional-dependencies[test]
- Revert manual type cast workarounds
- Ensures pydantic types from niconico.py are properly resolved during CI
- Simpler and more maintainable than type annotation workarounds

Commit:
Merge upstream/dev into add-niconico-provider

- Resolve conflict in audio.py by accepting upstream changes
- Update dependencies and other provider changes from upstream

Commit:
fix: remove deprecated enable_cache parameter from StreamDetails

- Remove enable_cache parameter removed in music-assistant-models v1.1.61
- This fixes mypy errors after merging upstream/dev changes

Commit:
nicovideo snapshots update

Commit:
refactor(nicovideo): implement fast HLS seeking with StreamType.CUSTOM

Replace StreamType.HTTP with CUSTOM for optimized seeking support.

Key improvements:
- Parse m3u8 once during stream conversion (not per seek)
- Generate dynamic m3u8 starting from seek segment
- Separate responsibilities: fetch → parse → generate → stream
- Use VERBOSE logging for seek operations
- Organize helpers into directory structure

Commit:
nicovideo update test,fixture,snapshot

Shi-553 committed:
Merge branch 'music-assistant:dev' into add-niconico-provider

Commit:
Remove FFmpeg submodule directory

Shi-553 committed:
Merge branch 'music-assistant:dev' into add-niconico-provider

Commit:
refactor: remove get_stream_details_for_mixin pattern, use normal override in TrackMixin

Commit:
docs: add detailed comments explaining StreamType.CUSTOM seek optimization

- Add comprehensive explanation in stream.py for why CUSTOM is used
- Document two-stage seek optimization (segment selection + input-side -ss)
- Explain input-side -ss limitation (can't cross segment boundaries)
- Add stage-by-stage comments in get_audio_stream and create_stream_context
- Clarify performance trade-off vs output-side -ss

Commit:
refactor(nicovideo): improve HLS streaming code structure and naming

Major changes:
- Rename NicovideoHLSProcessor → HLSSeekOptimizer for clarity
- Extract HLS data models to separate hls_models.py file
- Standardize m3u8 → playlist terminology throughout codebase
- Add HLS prefix to public API names to avoid confusion with track playlists
- Introduce DOMAND_BID_COOKIE_NAME constant for authentication
- Update method names for better semantic clarity:
  * _get_stream_data → _prepare_conversion_data
  * convert_by_stream_data → convert_from_conversion_data
  * _fetch_hls_m3u8_text → _fetch_media_playlist_text
  * _serve_m3u8 → _serve_hls_playlist

File structure:
- hls_processor.py → hls_seek_optimizer.py (renamed)
- hls_models.py (new): ParsedHLSPlaylist, HLSSegment

All tests and fixtures updated accordingly.

Commit:
refactor(nicovideo/tests): improve test module architecture

- Simplify TypeToConverterMappingRegistry (remove 3 unused methods)
- Extract all stabilization logic into FieldStabilizer class
- Add helper method to ConverterTestRunner for clarity
- Document FixtureProcessorProtocol design intent

Reduces complexity while maintaining type safety and auto-update features

Commit:
refactor(nicovideo/tests): improve naming and separate concerns

Renamed classes for clarity:
- TypeToConverterMapping → APIResponseConverterMapping
- FixtureManager → FixtureLoader
- FixtureDataGenerators → APIFixtureCollector
- FixtureDataSaver → FixtureSaver
- PathToTypeMapper → FixtureTypeMappingCollector + FixtureTypeMappingFileGenerator

Changes:
- Updated method names: generate_* → collect_* in APIFixtureCollector
- Fixed type completion by changing list to tuple for API_RESPONSE_CONVERTER_MAPPINGS
- Removed re-export pattern from fixtures/__init__.py
- Added comprehensive documentation to test_converters.py

Benefits: clearer responsibilities, better type inference, improved maintainability

marcelveldt commented:
> Or is there a design reason for always prioritizing album images?

Yes, this has come up a few times for consistency reasons people seem to always prefer the album image.

Can we flip this question, why is this provider an exception and should the track thumb be preferred ?

Commit:
fix: update niconico.py dependency to latest commit

Shi-553 commented:
In this provider, Niconico's "[Series](https://qa.nicovideo.jp/faq/show/14606?site_domain=default)" feature is mapped to albums.

### Why Track Thumbnails Should Be Preferred for Niconico

**1. Series/Album Structure Characteristics:**
- Adding videos to a series is **optional** for creators
- Many creators organize their videos into series by **genre** rather than by cohesive album concepts
  - Examples: "Original Songs", "Remixes", "Song Covers", etc.
- This results in **albums containing many unrelated tracks**, reducing visual recognition

**2. Track Thumbnails Are Essential Identity:**
- Niconico **requires** a thumbnail for every video upload
- The thumbnail is a **core part of each track's identity** for users
- Niconico users are accustomed to identifying content primarily by these track thumbnails

**3. User Experience Impact:**
- Prioritizing album images would be **confusing for Niconico users** who expect to see individual track thumbnails
- The current behavior breaks the familiar Niconico UX pattern

### Alternative Approach

If modifying the core behavior is not desirable, **I would like to prevent this provider from using album images entirely** to maintain consistency with the Niconico platform's UX expectations.

Shi-553 committed:
Merge branch 'music-assistant:dev' into add-niconico-provider

marcelveldt commented:
> If modifying the core behavior is not desirable, I would like to prevent this provider from using album images entirely to maintain consistency with the Niconico platform's UX expectations

For now, I can not fully oversee the implications of modifying the core behavior but I will change it anyways - we can use the beta period to collect user feedback and if some unexpected side effect is reported, we will need to find a way to specify the behavior fine grained per provider.

marcelveldt commented on .github/workflows/nicovideo-fixtures-update-and-test.yml:
please remove this from the PR, if you need to do specific CI tasks and testing of your code for translating api to python models, that needs to be in a dedicated repository  for your client library.

Shi-553 replied:
Thanks for the feedback. Removed the CI workflow and all fixture generation infrastructure from this PR.

Only test execution code and static fixtures remain.

marcelveldt commented on .vscode/launch.json:
please limit your PR to you provider only. 
Adding this quick launch command does however make a lot of sense, it just belongs in a dedicated PR

Shi-553 replied:
Removed the VS Code launch configuration changes.

marcelveldt commented on music_assistant/providers/nicovideo/helpers/utils.py:
```diff
@@ -0,0 +1,93 @@
+        cache_key,
+        track.to_dict(),
+        provider=provider.lookup_key,
+    )
```
Note that we recently did a bunch of changes to caching, basically we removed "automagic" caching by default from the core controllers making it a responsibility of the providers - which know better what and how long to cache.

Also we added support for a cache category to more fine grained split up the results in the cache (and it will give a bit of s speed boost due the indexes)

Shi-553 replied:
Migrated to the new @use_cache decorator system.

Applied cache durations based on SoundCloud patterns:
- get_artist/track/playlist: 14 days
- search: 3 hours

marcelveldt commented on music_assistant/providers/nicovideo/provider_mixins/artist.py:
```diff
@@ -0,0 +1,65 @@
+            async for track in self.mass.music.tracks.iter_library_items(
+                provider=self.instance_id,
+            ):
+                for artist in track.artists:
```
not sure I like this mixup of iterating the tracks controller from within a provider.
You should only return in-library or favorited artists here, nothing more, nothing less. If your provider does not have a concept of library artists, just leave out the feature to fallback to the library database of collecting this for you.

Shi-553 replied:
Fixed by removing the tracks controller iteration. The method now only returns followed artists from Niconico's native following system, which represents the actual in-library artists for this provider.

Also removed get_library_tracks and get_library_albums methods that had similar issues, along with their corresponding ProviderFeature flags.

marcelveldt commented on music_assistant/providers/nicovideo/provider_mixins/artist.py:
```diff
@@ -0,0 +1,65 @@
+                        yield self.mass.music.artists.artist_from_item_mapping(artist)
+
+        # Include followed artists if user is logged in
+        following_artists = await self.service_manager.user.get_own_followings()
+        for artist in following_artists:
+            yield artist
```
yeah, this is the only part you should return, scrap the part above

marcelveldt commented on tests/providers/nicovideo/fixtures/scripts/api_fixture_collector.py:
tbh I think all the fixture creating stuff needs to exist in a dedicated (client library) repository to split up responsibilities.

Shi-553 replied:
Created dedicated repository: **[music-assistant-nicovideo-fixtures](https://github.com/Shi-553/music-assistant-nicovideo-fixtures)**

**Separation:**
- **Fixture generation repo**: API collection, processing, type mapping generation
- **This PR**: Static fixtures + test execution only

**Update process**: Generate fixtures in dedicated repo → copy to `tests/providers/nicovideo/fixture_data`

This keeps generation logic and credentials out of the main codebase.

Commit:
Refactor: Reorganize fixture structure and update documentation

- Move fixture generation to separate repository (music_assistant_nicovideo_fixtures)
- Restructure fixtures: use fixture_data/ for generated data
- Update fixture loading to use new structure with fixture_type_mappings
- Remove embedded fixture generation scripts from tests
- Add comprehensive README for test directory
- Update VS Code tasks for easier fixture updates
- Delete obsolete GitHub workflow for fixture generation

Commit:
refactor: apply @use_cache to nicovideo provider methods

- Apply @use_cache decorator to get_* methods for improved performance
- Remove manual cache handling (cache_track helper and complex album info updates)
- Adjust cache durations based on SoundCloud patterns:
  - get_artist: 14 days (was 30 days)
  - get_track: 14 days (was 7 days)
  - get_playlist: 14 days (was 30 days)
  - get_playlist_tracks: 3 hours (unchanged)
  - search: 3 hours (was 1 hour)
  - get_album methods: 7 days (unchanged)
- Simplify code by leveraging Music Assistant's standard caching system
- Fix imports and add proper noqa comments for @use_cache usage

Commit:
revert: remove VS Code launch configuration changes

Remove the 'Music Assistant: Snapshot Update' launch configuration that was
added in previous commits. As requested in PR review, this change belongs
in a separate PR focused on development tools rather than the nicovideo
provider implementation.

Commit:
refactor: remove inappropriate cross-controller iteration

Address PR feedback by removing methods that iterate other controllers:
- Remove get_library_tracks and get_library_albums methods
- Remove library_add_for_mixin and library_remove_for_mixin methods
- Remove get_library_artists iteration over tracks controller
- Remove corresponding ProviderFeature flags from SUPPORTED_FEATURES
- Fix imports and TYPE_CHECKING blocks

The provider now only handles its native concepts (followed artists,
mylists, series) without inappropriately accessing other controllers.
This aligns with Music Assistant's architecture where providers should
only return in-library or favorited items from their own domain.

Commit:
fix: remove unintended cache category from get_track

Remove category=1 parameter from @use_cache decorator in get_track method
to maintain consistency with other cached methods that use default
category 0. This aligns with the pattern used across other providers
and removes unnecessary complexity.

Commit:
fix: Move media item types out of TYPE_CHECKING to fix @use_cache runtime errors

- Move Album, Artist, Track imports from TYPE_CHECKING to regular imports in artist.py
- Add noqa: TC002 comments for @use_cache decorator compatibility
- Format media_items imports in track.py consistently
- Fixes NameError: name 'Artist' is not defined when using @use_cache decorator
- Aligns with other Music Assistant provider patterns

Commit:
Merge remote-tracking branch 'upstream/dev' into add-niconico-provider

Commit:
refactor: improve get_following_activities and fix throttler settings

- Call get_track directly instead of get_provider_item for simplicity
- Remove unnecessary semaphore (throttler already handles rate limiting)
- Fix throttler settings: rate_limit=5, period=1 (5 req/sec) for high priority
- Fix low priority throttler: period=1 for consistency

Commit:
refactor: optimize get_following_activities to use direct Activity conversion

- Add convert_by_activity() method to track converter for lightweight feed conversion
- Replace get_track() loop with direct Activity-to-Track conversion
- Remove unnecessary asyncio.gather overhead
- Use ItemMapping for artist info instead of creating Owner objects
- Improves performance by reducing API calls from N+1 to 1

Commit:
Merge remote-tracking branch 'upstream/dev' into add-niconico-provider

# Conflicts:
#	pyproject.toml

Commit:
Update test snapshots for upstream StreamDetails changes

marcelveldt commented on pyproject.toml:
```diff
@@ -62,14 +62,15 @@ test = [
   "syrupy==5.0.0",
   "tomli==2.3.0",
   "ruff==0.14.3",
+  "niconico.py @ git+https://github.com/Shi-553/niconico.py@b2d781de11df4bf42d0d42037ac7f5abb6822fda",
```
ensure to change this to an pypi release version number

Commit:
Merge upstream/dev into add-niconico-provider

MarvinSchenkel commented:
@Shi-553 We're getting close to the 2.7 release. If you have the time, please adjust the last bit of feedback so we can merge this 🚀

Commit:
Merge upstream/dev into add-niconico-provider

Commit:
feat: switch from Git to PyPI reference for niconico.py-ma

- Update pyproject.toml to use niconico.py-ma==2.1.0.post1 from PyPI
- Update requirements_all.txt with PyPI package reference
- Update nicovideo provider manifest.json
- Update fixtures project dependency

This addresses PR feedback to use PyPI packages instead of Git references.

Commit:
chore: update fixture data docstring from generator

- Minor docstring update from fixture generation process
- No functional changes

Shi-553 commented:
Thanks for the heads up on the 2.7 timeline!

I've addressed the PyPI feedback by publishing the package and updating all references:

**PyPI Package**: https://pypi.org/project/niconico.py-ma/2.1.0.post1/
- Published as `niconico.py-ma` to avoid conflicts with the original package
- Updated all Git references

Note: This is a temporary fork package. Once the upstream PR to the original niconico.py is merged, we plan to switch back to the official PyPI release to maintain consistency with the upstream project.


Let me know if anything else is needed!

MarvinSchenkel commented on pyproject.toml:
```diff
@@ -61,14 +61,15 @@ test = [
   "syrupy==5.0.0",
   "tomli==2.3.0",
   "ruff==0.14.4",
+  "niconico.py-ma==2.1.0.post1",
```
This should not be required as providing it in the `manifest.json` is enough for the dependency to be installed.

Shi-553 replied:
Removed.
I had added it previously because the automated tests were failing, but if specifying it in manifest.json is sufficient for dependency installation, it's no longer needed.

Commit:
refactor(nicovideo): address Copilot review feedback

- Simplify to_dict_for_snapshot to assert dict return type
- Switch auth service logging to lazy % formatting
- Fix double space in recommendation folder label
- Document 'additionals' in codespell ignore list

MarvinSchenkel commented:
Also have a look at the copilot suggestions, there are some valid points there.

Shi-553 commented on tests/providers/nicovideo/helpers.py:
```diff
@@ -0,0 +1,72 @@
+        return sorted_result
+    else:
+        # This should not happen given the input, but satisfies mypy
+        return item_dict
```
I’ve simplified this helper to assert that the result is a dict and return it directly, removing the unreachable fallback branch.

Shi-553 commented on music_assistant/providers/nicovideo/services/auth.py:
```diff
@@ -0,0 +1,148 @@
+                        return False
+        self.logger.error(
+            f"Could not login after exceeding the maximum number of retries ({max_retries})."
+        )
```
Updated this warning to use lazy % formatting so the message is only formatted when the log line is emitted.

Shi-553 commented on music_assistant/providers/nicovideo/provider_mixins/explorer.py:
```diff
@@ -0,0 +1,123 @@
+            recommendation_folders.append(
+                RecommendationFolder(
+                    item_id="nicovideo_history",
+                    name="Recently watched  (nicovideo history)",
```
Fixed the extra space so the label now reads “Recently watched (nicovideo history)”.

Shi-553 commented on pyproject.toml:
```diff
@@ -61,14 +61,15 @@ test = [
 [tool.codespell]
 # explicit is misspelled in the iTunes API
-ignore-words-list = "provid,hass,followings,childs,explict,commitish,"
+ignore-words-list = "provid,hass,followings,childs,explict,additionals,commitish,"
```
“additionals” here is a fixed field name coming from the Nicovideo API responses and our recorded fixtures (e.g. the `additionals` object in the search JSON).
Renaming it would diverge from the real API, so I kept it in the ignore list and added a comment to document that it’s an intentional protocol field name.

Shi-553 commented on music_assistant/providers/nicovideo/services/auth.py:
```diff
@@ -0,0 +1,148 @@
+
+from __future__ import annotations
+
+import asyncio
```
The top-level `import asyncio` is used at runtime (e.g. `asyncio.to_thread`, `asyncio.sleep`, `asyncio.CancelledError`), while `from asyncio import TimerHandle` is under `TYPE_CHECKING` and only used for type hints.
That import is stripped at runtime, so this doesn’t affect actual behavior.
I’ll keep this structure for clearer type annotations, but can switch to `asyncio.TimerHandle` everywhere if the project prefers.

Commit:
chore: remove niconico.py-ma dependency from test requirements

Shi-553 commented on music_assistant/providers/nicovideo/services/auth.py:
```diff
@@ -0,0 +1,148 @@
+                        self.logger.warning(
+                            f"Network or DNS error occurred: {e}. "
+                            f"Retrying in {retry_delay_seconds} seconds..."
+                        )
```
Updated this warning to use lazy % formatting so the message is only formatted when the log line is emitted.

Commit:
Merge remote-tracking branch 'upstream/dev' into add-niconico-provider

Commit:
Reimplement get_own_followings to fetch followed artists

Restore the ability to retrieve users that the current user is following,
which was removed during config simplification.

Commit:
Remove unused user service methods

Remove methods that are no longer used in the current MA architecture:
- get_following_mylists() - empty stub returning []
- get_following_playlists() - only called get_following_mylists
- follow_user() - no MA interface to call this
- unfollow_user() - no MA interface to call this

These were remnants from config simplification that have no callers.

Commit:
Remove unused methods from nicovideo implementation

- Remove NicovideoUserService.get_own_videos()
- Remove NicovideoVideoService.like_video()
- Remove NicovideoSeriesService.get_own_series()
- Remove NicovideoSearchService.search_videos_by_tag()
- Remove unused imports (VideoSearchSortKey, VideoSearchSortOrder)

MarvinSchenkel commented on music_assistant/providers/nicovideo/services/manager.py:
```diff
@@ -0,0 +1,193 @@
+        else:
+            self.logger.warning("Error %s called from %s: %s", operation, caller_info, err)
+
+    async def _call_with_throttler[T, **P](
```
We already have a helper for throttling in helpers/throttle_retry. Any change that one can be used instead of defining a new one?

Shi-553 replied:
We're already using the `ThrottlerManager` from `helpers/throttle_retry`. The wrapper method `_call_with_throttler` exists because niconico.py provides synchronous functions, so we use `asyncio.to_thread()` to run them in a thread pool while maintaining proper throttling.

However, your review prompted us to simplify the implementation. We removed:
- Unused priority control (LOW priority throttler was defined but never used)
- Excessive logging (argument sanitization, verbose success messages)

Change: https://github.com/music-assistant/server/pull/2339/commits/11e4b3d7eb7a6415503a93b9cc2c26280b3e5b9a

MarvinSchenkel commented on music_assistant/providers/nicovideo/helpers/hls_models.py:
```diff
@@ -0,0 +1,95 @@
+"""HLS data models and parsing for nicovideo provider."""
```
There are already HLS helpers available under helpers/playlist.py. Any chance we can use those over defining a new one?

Shi-553 replied:
Thank you for the suggestion. I've investigated whether we can use the existing playlists.py.

**Conclusion:** We cannot directly use `parse_m3u()` as-is, but the HLS manipulation logic is generic and could be moved to playlists.py.

**Why `parse_m3u()` doesn't fit:**
- `parse_m3u()` extracts playlist items (returns `list[PlaylistItem]`)
- nicovideo needs to preserve and reconstruct the entire HLS structure (headers, init segment, encryption key, segments)
- Specifically, `_generate_dynamic_playlist()` in hls_seek_optimizer.py reconstructs HLS playlists starting from a specific segment for fast seeking

**However, the HLS logic is provider-agnostic:**
- `ParsedHLSPlaylist` and `HLSSegment` models contain no nicovideo-specific code
- `_calculate_start_segment()` and `_generate_dynamic_playlist()` are pure HLS manipulation
- Only the FFmpeg header generation (User-Agent, authentication cookies) is nicovideo-specific

**Potential refactoring:**
These generic HLS utilities could be moved to playlists.py to benefit other providers. 
Would you like me to create a follow-up PR for generalizing the HLS utilities?

MarvinSchenkel replied:
Yep let's do this in a follow-up PR!

MarvinSchenkel commented:
Last two questions from my side and then we can merge this! 👏

Commit:
refactor: simplify throttler implementation in nicovideo provider

- Use existing helpers/throttle_retry.ThrottlerManager instead of custom implementation
- Remove unused priority control (ApiPriority.LOW was never used)
- Remove verbose logging (sanitization, argument summaries, success logs)
- Keep only error logging with caller info for diagnostics
- Align with other providers' logging patterns (Spotify, Qobuz, Tidal)

Commit:
docs: Add explanation for 'Domand' spelling in niconico API

Clarify that 'Domand' is the actual spelling used in niconico's API
(not a typo). This naming appears in API endpoints and throughout
their media delivery system classes.

Shi-553 commented on tests/providers/nicovideo/fixture_data/shared_types.py:
```diff
@@ -0,0 +1,28 @@
+from __future__ import annotations
+
+# Pydantic requires runtime type information, so these imports cannot be in TYPE_CHECKING block
+from niconico.objects.video.watch import WatchData, WatchMediaDomandAudio  # noqa: TC002
```
The word "Domand" is actually the correct spelling used in niconico's API, not a typo. This naming appears consistently throughout their infrastructure:

- API endpoints: `https://asset.domand.nicovideo.jp/`
- Data model classes: `WatchMediaDomand`, `WatchMediaDomandVideo`, `WatchMediaDomandAudio`
- Cookie names: `domand_bid`

I've added a clarifying comment in the constants file where `DOMAND_BID_COOKIE_NAME` is defined, as this constant is imported and used across multiple files in the provider.

Commit:
docs(nicovideo): clarify HLS seeking limitation in stream.py

Improve technical accuracy of comment explaining why input-side -ss
fails with HLS without playlist reconstruction.

Previous comment incorrectly suggested the issue was about 'crossing
segment boundaries', but the actual problem is that FFmpeg cannot
identify target segments before parsing the playlist structure.

This clarification better reflects the implementation constraint.

MarvinSchenkel reviewed (approved):
Thanks @Shi-553 , great work!

OzGav commented:
@Shi-553 I have created a docs page based on what I can see in this PR and a quick look at the provider config. If you could review it and let me know of any errors or additions. We would normally explain to people, for example, how to get the cookie.

https://github.com/music-assistant/beta.music-assistant.io/pull/56
